mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/task/SolveMessageDeletedInconsistenciesService.java

```diff
@@ -0,0 +1,103 @@
+
+import org.apache.james.mailbox.cassandra.ids.CassandraId;
+import org.apache.james.mailbox.cassandra.mail.CassandraDeletedMessageDAO;
+import org.apache.james.mailbox.cassandra.mail.CassandraMailboxDAO;
+import org.apache.james.mailbox.cassandra.mail.CassandraMessageIdDAO;
+import org.apache.james.mailbox.model.MessageRange;
+import org.apache.james.task.Task;
+import org.apache.james.util.streams.Limit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+
+import reactor.core.publisher.Mono;
+
+public class SolveMessageDeletedInconsistenciesService {
+
+    public record Context(AtomicLong processedMailboxEntries,
+                          ConcurrentLinkedDeque<CassandraId> errors) {
+
+        record Snapshot(long processedMailboxEntries, ImmutableList<CassandraId> errors) {
+        }
+
+        public Context() {
+            this(new AtomicLong(0), new ConcurrentLinkedDeque<>());
+        }
+
+        void incrementProcessedMailboxEntries() {
+            processedMailboxEntries.incrementAndGet();
+        }
+
+        void addError(CassandraId cassandraId) {
+            errors.add(cassandraId);
+        }
+
+        Snapshot snapshot() {
+            return new Snapshot(processedMailboxEntries.get(), ImmutableList.copyOf(errors));
+        }
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SolveMessageDeletedInconsistenciesService.class);
+    private final CassandraMessageIdDAO messageIdDAO;
+    private final CassandraDeletedMessageDAO deletedMessageDAO;
+    private final CassandraMailboxDAO mailboxDAO;
+
+    @Inject
+    public SolveMessageDeletedInconsistenciesService(CassandraMessageIdDAO messageIdDAO,
+                                                     CassandraDeletedMessageDAO deletedMessageDAO,
+                                                     CassandraMailboxDAO mailboxDAO) {
+        this.messageIdDAO = messageIdDAO;
+        this.deletedMessageDAO = deletedMessageDAO;
+        this.mailboxDAO = mailboxDAO;
+    }
+
+    public Mono<Task.Result> fixInconsistencies(Context context) {
+        return mailboxDAO.retrieveAllMailboxes()
+            .concatMap(mb -> fixDeletedMessagesInconsistencyPerMailbox((CassandraId) mb.getMailboxId(), context)
+                .doOnNext(any -> context.incrementProcessedMailboxEntries()))
+            .reduce(Task.Result.COMPLETED, Task::combine);
+    }
+
+    private Mono<Task.Result> fixDeletedMessagesInconsistencyPerMailbox(CassandraId cassandraId, Context context) {
+        return deletedMessageDAO.removeAll(cassandraId)
+            .then(messageIdDAO.retrieveMessages(cassandraId, MessageRange.all(), Limit.unlimited())
+                .filter(metadata -> metadata.getComposedMessageId().getFlags().contains(Flags.Flag.DELETED))
+                .collect(Collectors.mapping(metadata -> metadata.getComposedMessageId().getComposedMessageId().getUid(), Collectors.toList()))
```

Comment:
Why not 

```
xxx.map(...)
   .collect(ImmutableList.toImmutableList())
```

?

Reply:
i think it is similar, just another syntax?

Reply:
This proposal is both more verbose and not a common practice. As far as I know we never used it in our code.

It is not because we can that we should...